lib/pbench/server/database/models/datasets.py

```diff
@@ -857,21 +846,44 @@ class Metadata(Database.Base):
     # {"dashboard.seen": True}
     DASHBOARD = "dashboard.*"
 
-    # USER is arbitrary data saved on behalf of the owning user, as a JSON
-    # document.
+    # USER_FAVORITE allows direct access (e.g., in /datasets/list) to the
+    # user-specific "favorite" metadata. NOTE: for now, this is hardcoded as
+    # False until we have a framework for user-specific metadata.
     #
-    # Note that the hierarchical key management in the getvalue and setvalue
-    # static methods (used consistently by the API layer) allow client code
-    # to interact with these either as a complete JSON document ("user") or
-    # as a full dotted key path ("user.contact.name.first") or at any JSON
-    # layer between.
+    # {"user.favorite": False}
+    USER_FAVORITE = "user.favorite"
+
+    # USER_ACCESS allows direct access (e.g., in /datasets/list) to the set of
+    # access policies the authenticated user is allowed to set. Initially this
+    # is always just private and public; in the future this can be extended to
+    # support groups.
     #
-    # API keyword validation uses the trailing ".*" here to indicate that only
-    # the first element of the path should be validated, allowing the client
-    # a completely uninterpreted namespace below that.
+    # {"user.access": ["private", "public"]}
+    USER_ACCESS = "user.access"
```

Comment:
It's the set of "access" values the specific authenticated user is allowed to set. So, yeah, it's a server IAM setting specific to the authenticated user. As such I like it under a "user" namespace, whereas under the "server" namespace it'd need a more specific name. Or maybe it needs a more specific name here, anyway. Code is easy: names are hard.

Sigh; but, then again, this won't vary among the datasets returned as `user.favorite` will. I put this here because we *have* a mechanism for dataset metadata, and I like the idea of making user favorite data visible when querying a list of datasets instead of requiring a separate user metadata call for each dataset. (Plus of course the fact that we don't currently have any such mechanism.)

But you're probably right: I shouldn't cram this in here, even though it was convenient.

Reply:
Right:  the value of this key is not associated with any given dataset...but, then again, in the previous model, nothing in the `user` namespace was associated with a given dataset.  I would put it under the `server` namespace, for the reason that you point out; and, there, it would potentially return the server's value associated with the requestor's login.

As for mechanism, I though Nikhil implemented most of what we needed...it just didn't have sufficient priority at the time to be integrated...perhaps that has finally changed!

Reply:
This is an IAM configuration that should be accessible in some form through the user profile. Maybe it even makes sense to add that now, as a read-only `User` object property, which current would be hardcoded to the `["public", "private"]` list and can be extended in the future assuming the entire user DB model doesn't change with Keycloak.

My intent here was to provide a shortcut where the dashboard might need the information, as part of querying a dataset, to populate a PUBLISH pulldown. The repetition of having the same user/IAM-defined values for each dataset in a list does bug me a bit, although it's convenient.

I think it's conceivable that we could have dataset-specific lists. E.g., if a dataset was owned by a group, an authorized user might be able to make the data public, or choose another related group, but maybe couldn't make it "private". If we went that way (which sounds too confusing to be viable), putting it here makes sense. Since that's unlikely, though, I guess adding it to the `User` is probably a better choice.

Reply:
My instinct is that the list of operations (e.g., "publish") and the list of permissions (e.g, "read", "write", "delete") should be basically orthogonal -- when the user makes a request, the server checks their intersection, and if there's not a "true" there, then the user gets a `403`.  I don't think that we need to bother trying to populate pull downs or to gray out options on a per-dataset basis.  That seems like it will be a lot of effort (or data, or bandwidth, or...) for very little return.  (That is, is it really better to say "you cannot do that" (by removing or graying out the option) than it is to let the user try and _then_ say "you cannot do that"?)